### ISSUE TITLE
Distributed Tracing WG: Baggage specification

### ISSUE THREAD

**Issue by SergeyKanzhelev** [w3ctag/design-reviews#650]
Ya ya yawm TAG!

I'm requesting a TAG review of Baggage specification

Baggage is a header format and behavior specification that enables scenarios of opaque values propagation along the distributed trace. Distributed Trace can spawn across multiple components and types of servers: browsers, proxies and
load balancers, server applications, and SaaS services. Baggage from incoming request is propagated to all related outgoing calls. This enables many inter-operability scenarios for end users when monitoring and troubleshooting highly distributed applications. There is a lot of prior art in a form of vendor-specific headers, this specification solves the interoperability issue when many app components may not be controlled by the end user.

  - Explainer (minimally containing user needs and example code): https://github.com/w3c/distributed-tracing-wg/blob/main/EXPLAINER.md
  - Specification URL: https://w3c.github.io/baggage/
  - Tests: WIP
  - Security and Privacy self-review²: WIP
  - GitHub repo (if you prefer feedback filed there): https://github.com/w3c/baggage/
  - Primary contacts (and their relationship to the specification):
      - [Sergey Kanzhelev](https://github.com/SergeyKanzhelev), [Google], editor
      - [Daniel Khan](https://github.com/danielkhan), [Dynatrace], editor 
  - Organization(s)/project(s) driving the specification: 
  - Key pieces of existing multi-stakeholder review or discussion of this specification:
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status): https://github.com/w3c/baggage/issues

Further details:

  - [X] I have reviewed the TAG's [Web Platform Design Principles](https://w3ctag.github.io/design-principles/)
  - Relevant time constraints or deadlines: we want to get to CR this calendar year
  - The group where the work on this specification is currently being done: [Distributed Tracing Working Group](https://www.w3.org/groups/wg/distributed-tracing)
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): [Distributed Tracing Working Group](https://www.w3.org/groups/wg/distributed-tracing)
  - Major unresolved issues with or opposition to this specification: none
  - This work is being funded by: multiple companies

You should also know that...

Previous form of this specification is built-in into [ASP.NET framework](https://devblogs.microsoft.com/aspnet/improvements-in-net-core-3-0-for-troubleshooting-and-monitoring-distributed-apps/#passing-additional-context), current version of specification is implemented by many [OpenTelemetry](https://opentelemetry.io/) SDKs.

We found that it is great to have predictability in which field is being automatically propagated as it increases transparency for end-user, as oppose to numerous vendor-specific headers.

We'd prefer the TAG provide feedback as (please delete all but the desired option):

  🐛 open issues in our GitHub repo for **each point of feedback**

**Comment by otherdaniel** [comment]
Can you clarify what a user-agent is meant to do on receiving this header, if anything?

This issue came up during Chromium Security & Privacy review of this issue, and we were unclear whether the spec is meant to apply to browsers / user agents or not. The spec seems primarily concerned with "downstream services", but the last example in the explainer shows a web browser echoing a "W3C Trace ID" back to a server, which would make this functionality cookie-equivalent. However, the implications of being cookie-like (life time? origin bound? which permissions apply?) do not seem to be spelled out clearly.

**Comment by dyladan** [comment]
> Can you clarify what a user-agent is meant to do on receiving this header, if anything?

For the baggage header, the user-agent will not do anything other than send the header. In version 1, there is no response header and we have no plans to introduce one in future versions at the moment.

> This issue came up during Chromium Security & Privacy review of this issue, and we were unclear whether the spec is meant to apply to browsers / user agents or not.

Other than sending the baggage header just like any other header, the user-agent has no responsibilities. It is conceivable that a user-agent would want to add baggage entries, but that is not part of the specification.

>The spec seems primarily concerned with "downstream services", but the last example in the explainer shows a web browser echoing a "W3C Trace ID" back to a server, which would make this functionality cookie-equivalent. However, the implications of being cookie-like (life time? origin bound? which permissions apply?) do not seem to be spelled out clearly.

The "W3C Trace ID" is part of the separate but related [W3C Trace Context](https://www.w3.org/TR/trace-context/) specification which is not a part of the baggage specification which we are asking to be reviewed.

**Comment by arturjanc** [comment]
> For the baggage header, the user-agent will not do anything other than send the header.

This on its own is concerning security-wise and should be covered in the spec and explainer. For example, what happens on cross-origin requests (e.g. those made in `no-cors` mode): would the baggage header be included? How should cross-origin redirects be handled?  

I think a conceptually simple way to address this could be to guarantee that the baggage header can only be attached on requests that are same-origin to the response that included the baggage information. The spec already mentions that baggage data "does not leak beyond defined trust boundaries" but my guess is that the folks working on this need to put a little more thought into how this would work on the web.

**Comment by torgo** [comment]
Hi @SergeyKanzhelev - thanks for sending us this. We're still looking for some more info related to security & privacy - specifically the response to the questionnaire... Also can you be a bit more clear on how you "forbid encoding user identifiable data"?

**Comment by torgo** [comment]
Also can you please let us know the multistakeholder status on this? What is the implementation status across other engines/browsers?

**Comment by samuelweiler** [comment]
@torgo, thank you for drawing this to PING's attention by adding a privacy-tracker label.  

As a reminder for the WG, when you're ready for a PING review, please follow the process at https://www.w3.org/Guide/documentreview/#how_to_get_horizontal_review

**Comment by SergeyKanzhelev** [comment]
Sorry for the long delay in replies.

> > For the baggage header, the user-agent will not do anything other than send the header.
> 
> This on its own is concerning security-wise and should be covered in the spec and explainer. For example, what happens on cross-origin requests (e.g. those made in `no-cors` mode): would the baggage header be included? How should cross-origin redirects be handled?

At the moment there are no plans to implement baggage as part of the user-agent. It will typically be done in the JavaScript library executing on the page. This JavaScript library will need to be configured to account for cors correctly.

> I think a conceptually simple way to address this could be to guarantee that the baggage header can only be attached on requests that are same-origin to the response that included the baggage information. The spec already mentions that baggage data "does not leak beyond defined trust boundaries" but my guess is that the folks working on this need to put a little more thought into how this would work on the web.

We believe that defining the trust boundaries is not the part of specification, thus no specific details on how to implement cors or configure trust boundaries in the web. Trust boundaries, especially involving many components of a distributed application, are typically very specific to the application and cannot be easily generalized as a specification. This, however, may be a whole new topic for the distributed tracing working group for the future exploration.



> We're still looking for some more info related to security & privacy - specifically the response to the questionnaire... Also can you be a bit more clear on how you "forbid encoding user identifiable data"?

I think this is referring to the privacy section of a specification. Specification does not solve the problem of what data will be propagated thru the distributed application components. It solves the problem of standardizing the way data is propagated so centralized solutions may be developed to inspect and secure this propagation. Many tools today allow to propagate this information using custom headers which are vendor specific and not clearly documented. So users of these tools have less control over the data being transmitted. When tools will follow the spec, it will be easier for user to control what is transmitted.

**Comment by SergeyKanzhelev** [comment]
> I didn't see any reference to [Structured Headers](https://datatracker.ietf.org/doc/html/rfc8941), can your header format be expressed this way?

We had an extensive discussions on whether to use structured headers for this or not. For multiple reasons we decided to proceed with the cookies-like format and not support strongly typed values. I cannot find the rationale clearly documented, I will take this as an action item

**Comment by hadleybeeman** [comment]
Hi @SergeyKanzhelev. We are looking at this in our W3CTAG breakout, and a couple of questions have come up.

1. Who is the user for this? What need of theirs is it meeting? It would help us a lot of if you could put those points in your explainer.

2. We are concerned about the privacy/security implications of opening up a metadata channel that the user can't control. Have you thought through any attack scenarios? And if so, what are your thoughts on how to make them less likely to happen?

The [security](https://w3c.github.io/baggage/#security-considerations) and [privacy](https://w3c.github.io/baggage/#privacy-considerations) sections of your spec talk about the responsibilities of application owners and systems. Does your approach let users protect themselves too? Or might it be easier to track them with your proposal?

We'd be grateful for your thoughts on those. Thanks!

**Comment by arturjanc** [comment]
> At the moment there are no plans to implement baggage as part of the user-agent. It will typically be done in the JavaScript library executing on the page. This JavaScript library will need to be configured to account for cors correctly.

I'm having a bit of difficulty reconciling this with the description in the [explainer](https://github.com/w3c/distributed-tracing-wg/blob/main/EXPLAINER.md): _"We also propose a response header which can be used to report a trace ID back to the caller to ... correlate the initial page load of a browser and all subsequent requests to a server side trace"_ and _"The browser uses this information to provide a trace ID for all subsequent requests within this one load cycle"_.

From your explanation above I understand that the user's browser would simply ignore the header and the responsibility of propagating the baggage value would be on the client-side JS executing on a given page. This brings up a few thoughts:
1. How will scripts executing on a page learn the value of the `baggage` header? Are you also proposing a JS API to get this data?
2. This means that the baggage will -- in practice -- be propagated only for a subset of requests made by a given page, and will exclude `no-cors` resource loads. I think this means that the value will only be propagated for programmatic loads using the `fetch()` API if the caller makes sure to add the `baggage` header.
3. Pages with some third-party resources that don't opt in via `Access-Control-Allow-Headers` may have a hard time enabling this. My guess is that most pages are in this situation.

**Comment by dyladan** [comment]
Hello TAG review team. This is just a friendly ping to see what the current status of this review is. Please let me know if you're waiting on anything from us (Distributed Tracing WG) at this point.

**Comment by plinss** [comment]
@dyladan thanks for the ping, we’ll get back to this next week

**Comment by kenchris** [comment]
I see that you currently support key/value pairs and properties. Do you expect this to cover all your use-cases? Have you looked at structured headers:? https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-header-structure-15

**Comment by torgo** [comment]
We're also missing info on multi-stakeholder support. I can't find this on Chrome Status.  @dyladan @SergeyKanzhelev can you comment?

**Comment by hadleybeeman** [comment]
Thanks for the thorough reply to my earlier questions, @SergeyKanzhelev! They were helpful. For what it's worth though, I still didn't really get the use case you're trying to support until I looked at the [working group's charter](https://www.w3.org/2020/03/proposed-distributed-tracing-wg.html). You might want to take a bit of the scope section from the charter to set the scene in your explainer.

Also, we've been talking about the name of the feature. `Baggage` is definitely a US/UK idiom, and may not be intuitive to anyone who hasn't been focusing on distributed tracing. Might it be useful to use something like `Trace ID additional header`? (Maybe best to do some user research on this — the important thing is that it's immediately understandable to as many of your users as possible.)

And finally, in the privacy vein, we talked about whether it would be good for this to behave differently in private/incognito browsing mode. It struck us that there is nothing to stop the application provider from sending headers derived from information tied to the machine or usual user (their IP address, their user ID, etc.). We wondered if it might make sense to ask user agents to ignore the trace ID altogether in when in private/incognito mode, so that the user's experience isn't tied to the experience that same user (or whoever they share a browser with) gets when not in private/incognito mode. What are your thoughts on this?

**Comment by ylafon** [comment]
Also, is there a reason why, in the [ABNF](https://w3c.github.io/baggage/#definition), you refer to rfc2616 and rfc7230? Why not using the same source (ideally the latest) for all the definitions?

**Comment by torgo** [comment]
In the [Privacy & Security Section](https://w3c.github.io/baggage/#privacy-of-the-baggage-header) it states "As such, the baggage header can contain user-identifiable data. Systems MUST ensure that the baggage header does not leak beyond defined trust boundaries and they MUST ensure that the channel that is used to transport potentially user-identifiable data is secured." But the term "trust boundaries" is not itself defined in the document. Also maybe you should refernece [secure contexts](https://www.w3.org/TR/secure-contexts/) directly instead of just saying "secured" to be clear about what you mean by secured here?

We're also still waiting for a response on the issues raised by Hadley about incognito mode and naming.

**Comment by kalyanaj** [comment]
Regarding this comment: https://github.com/w3ctag/design-reviews/issues/650#issuecomment-969128971: we don't require any browser support for this specification as this is implemented in JavaScript (e.g. OpenTelemetry JS SDKs). OpenTelemetry SDKs across multiple languages implement Baggage specification.

**Comment by kalyanaj** [comment]
Hello TAG Review Team: Thanks for your review comments above, I want to consolidate the list of questions discussed above and the current status:

1. **Naming**: https://github.com/w3c/baggage/issues/83: Response tracked as part of this issue.
2. **Private Browsing mode**: https://github.com/w3c/baggage/issues/84: Response tracked as part of this issue.
3. **Multi-stakeholder support**: Covered as part of this response at https://github.com/w3ctag/design-reviews/issues/650#issuecomment-1040777563 
4. **Structured Headers**: Covered as part of this response at https://github.com/w3ctag/design-reviews/issues/650#issuecomment-909543300
5. **Reference to the right RFC** (https://github.com/w3ctag/design-reviews/issues/650#issuecomment-969290866):  We have addressed this feedback as part of https://github.com/w3c/baggage/pull/79.

Please let us know your thoughts on the above and if you have any additional comments/ questions.

**Comment by torgo** [comment]
Hi @kalyanaj - thank you so much for preparing that summary addressing all the issues that were raised. We're just coming back to this today at our face-to-face meeting. We're happy with the responses you've provided. We'd still like to see some additional justification re: the structured headers topic (documented in your spec). However we're broadly happy with this.
